File: android/app/src/main/java/com/laneshadow/ui/components/molecules/DiscoveryLoadingOverlay.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.components.atoms.Skeleton
import kotlinx.coroutines.delay

/**
 * DiscoveryLoadingOverlay molecule component
 *
 * Skeleton overlay for initial route discovery loading state.
 * Features 300ms debounce to prevent flash on fast loads.
 * Semi-transparent glassmorphic design (map visible behind).
 *
 * Following React Native wrapper patterns from react-native/components/discovery/discovery-loading-overlay.tsx
 *
 * @param visible Whether the overlay is visible
 * @param testId Test ID for UI testing
 */
@Composable
fun DiscoveryLoadingOverlay(
    visible: Boolean,
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // 300ms debounce to prevent flash on fast loads
    var debouncedVisible by remember { mutableStateOf(false) }

    LaunchedEffect(visible) {
        if (visible) {
            delay(300) // 300ms debounce
            debouncedVisible = true
        } else {
            // Immediately hide when data loads
            debouncedVisible = false
        }
    }

    // Early return when not visible (after debounce)
    if (!debouncedVisible) {
        Box(modifier = Modifier.testTag(testId ?: "discovery-loading-overlay"))
        return
    }

    // Semi-transparent surface background (80% opacity = 0.8f alpha)
    val overlayBackgroundColor = theme.colors.surface.default.copy(alpha = 0.8f)

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(overlayBackgroundColor)
            .testTag(testId ?: "discovery-loading-overlay"),
    ) {
        Column(
            modifier = Modifier.padding(
                top = theme.space.lg,
                bottom = theme.space.lg,
            ),
        ) {
            // Skeleton for filter bar area
            FilterBarSkeleton(testId = testId)

            // Skeleton for route pins area
            RoutePinsSkeleton(testId = testId)
        }
    }
}

/**
 * Filter bar skeleton component
 *
 * Shows skeleton chips for archetypes filter.
 *
 * @param testId Test ID for UI testing
 */
@Composable
private fun FilterBarSkeleton(
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    Column(
        modifier = Modifier
            .padding(horizontal = theme.space.lg)
            .padding(bottom = theme.space.md),
    ) {
        // Skeleton chips for archetypes (80-100dp wide, 32dp tall, rounded)
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
        ) {
            Skeleton(
                width = 80.dp,
                height = 32.dp,
                shape = theme.radius.full,
                testID = testId?.let { "$it-chip-1" } ?: "discovery-loading-overlay-chip-1",
            )
            Skeleton(
                width = 100.dp,
                height = 32.dp,
                shape = theme.radius.full,
                testID = testId?.let { "$it-chip-2" } ?: "discovery-loading-overlay-chip-2",
            )
            Skeleton(
                width = 90.dp,
                height = 32.dp,
                shape = theme.radius.full,
                testID = testId?.let { "$it-chip-3" } ?: "discovery-loading-overlay-chip-3",
            )
        }
    }
}

/**
 * Route pins skeleton component
 *
 * Shows skeleton avatars and labels simulating map pins in a scattered pattern.
 *
 * @param testId Test ID for UI testing
 */
@Composable
private fun RoutePinsSkeleton(
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(horizontal = theme.space.lg),
        verticalArrangement = Arrangement.spacedBy(theme.space.xl * 2),
    ) {
        // Pin row 1
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space.xl * 2),
        ) {
            PinSkeleton(
                labelWidth = 50.dp,
                testId = testId?.let { "$it-pin-1" } ?: "discovery-loading-overlay-pin-1",
            )
            PinSkeleton(
                labelWidth = 60.dp,
                testId = testId?.let { "$it-pin-2" } ?: "discovery-loading-overlay-pin-2",
            )
        }

        // Pin row 2
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space.xl * 3),
        ) {
            PinSkeleton(
                labelWidth = 45.dp,
                testId = testId?.let { "$it-pin-3" } ?: "discovery-loading-overlay-pin-3",
            )
            PinSkeleton(
                labelWidth = 55.dp,
                testId = testId?.let { "$it-pin-4" } ?: "discovery-loading-overlay-pin-4",
            )
        }

        // Pin row 3
        Row(
            horizontalArrangement = Arrangement.spacedBy(theme.space.xl * 2),
            modifier = Modifier.padding(start = theme.space.lg),
        ) {
            PinSkeleton(
                labelWidth = 48.dp,
                testId = testId?.let { "$it-pin-5" } ?: "discovery-loading-overlay-pin-5",
            )
        }
    }
}

/**
 * Pin skeleton component
 *
 * Avatar skeleton + small label skeleton simulating a map pin.
 *
 * @param labelWidth Width of the label skeleton
 * @param testId Test ID for UI testing
 */
@Composable
private fun PinSkeleton(
    labelWidth: Dp,
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    Column(
        verticalArrangement = Arrangement.spacedBy(theme.space.xs),
    ) {
        // SkeletonAvatar (40dp circle matching RN default size)
        Skeleton(
            width = 40.dp,
            height = 40.dp,
            shape = theme.radius.full,
            testID = testId?.let { "$it-avatar" } ?: "discovery-loading-overlay-pin-avatar",
        )

        // Small label skeleton
        Skeleton(
            width = labelWidth,
            height = 12.dp,
            shape = theme.radius.sm,
            testID = testId?.let { "$it-label" } ?: "discovery-loading-overlay-pin-label",
        )
    }
}

File: ios/LaneShadow/Views/Molecules/DiscoveryLoadingOverlay.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Discovery Loading Overlay Component

/**
 * Discovery loading overlay molecule component
 *
 * Full-screen overlay with skeleton placeholders for filter bar and route pins.
 * Shows during data loading with 300ms debounce to prevent flash on fast loads.
 * Immediately hides when visible becomes false (no debounce on hide).
 *
 * Following React Native component from react-native/components/discovery/discovery-loading-overlay.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surface.default` with 0.8 opacity (80%)
 * - Layout:
 *   - Full-screen ZStack with ignoresSafeArea
 *   - Filter bar: horizontal scroll with chip skeletons (80-100pt wide, 32pt tall)
 *   - Route pins: grid layout with avatar + label skeletons
 * - Reuses existing LSSkeleton atom component
 *
 * ## Parameters
 * - visible: Whether overlay should be shown (false returns EmptyView, hides immediately)
 * - testID: Optional test identifier for testing (default: "discovery-loading-overlay")
 */
public struct LSDiscoveryLoadingOverlay: View {
    @Environment(\.theme) private var theme

    private let visible: Bool
    private let testID: String

    @State private var showOverlay = false

    private let debounceDelay: TimeInterval = 0.3 // 300ms

    public init(
        visible: Bool,
        testID: String = "discovery-loading-overlay"
    ) {
        self.visible = visible
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        if !visible {
            // Immediately hide when visible becomes false
            EmptyView()
        } else {
            GeometryReader { _ in
                ZStack {
                    // Semi-transparent background (80% opacity)
                    theme.colors.surface.default.opacity(0.8)
                        .ignoresSafeArea()

                    // Loading content with debounce
                    VStack(spacing: 0) {
                        // Filter bar skeleton placeholders
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(spacing: theme.space.sm) {
                                ForEach(0..<5, id: \.self) { _ in
                                    filterChipSkeleton()
                                }
                            }
                            .padding(.horizontal, theme.space.md)
                            .padding(.vertical, theme.space.sm)
                        }

                        Spacer()

                        // Route pin skeleton placeholders
                        VStack(spacing: theme.space.lg) {
                            ForEach(0..<3, id: \.self) { _ in
                                routePinSkeleton()
                            }
                        }
                        .padding(.horizontal, theme.space.md)
                        .padding(.bottom, theme.space.xl)
                    }
                    .accessibilityIdentifier("\(testID)-content")
                }
            }
            .ignoresSafeArea()
            .task {
                // Debounce: wait 300ms before showing overlay
                try? await Task.sleep(nanoseconds: UInt64(debounceDelay * 1_000_000_000))
                showOverlay = true
            }
        }
    }

    // MARK: - Private Helpers

    private func filterChipSkeleton() -> some View {
        // Random width between 80-100pt, height 32pt, rounded
        let randomWidth = CGFloat(Double.random(in: 80...100))
        return LSSkeleton(
            width: randomWidth,
            height: 32,
            cornerRadius: theme.radius.full
        )
    }

    private func routePinSkeleton() -> some View {
        // Avatar skeleton + small label skeleton
        HStack(spacing: theme.space.sm) {
            LSSkeleton(
                width: theme.size.avatarDefault,
                height: theme.size.avatarDefault,
                cornerRadius: theme.radius.full
            )

            LSSkeleton(
                width: 80,
                height: 16,
                cornerRadius: theme.radius.sm
            )
        }
    }
}

// MARK: - Preview

#Preview("DiscoveryLoadingOverlay - Visible") {
    ZStack {
        Color.gray.opacity(0.3) // Simulate map background

        LSDiscoveryLoadingOverlay(
            visible: true
        )
    }
    .laneShadowTheme()
}

#Preview("DiscoveryLoadingOverlay - Hidden") {
    ZStack {
        Color.gray.opacity(0.3) // Simulate map background

        LSDiscoveryLoadingOverlay(
            visible: false
        )
    }
    .laneShadowTheme()
}

#Preview("DiscoveryLoadingOverlay - Custom testID") {
    ZStack {
        Color.gray.opacity(0.3) // Simulate map background

        LSDiscoveryLoadingOverlay(
            visible: true,
            testID: "custom-loading-overlay"
        )
    }
    .laneShadowTheme()
}
